### PR TITLE
[Spec-Decode] Add DynamicProposer for per-sequence dynamic speculative decoding

### PR DESCRIPTION
**Pull request by yang926**
## Purpose

This PR introduces a new speculative decoding method, `eagle_dynamic`, which enables adaptive adjustment of the number of speculative tokens (`k`). This addresses the idea of adaptive speculation discussed in issue dynamic SL in \#17984.

The existing speculative decoding methods in vLLM use a fixed `k` for all sequences. This can be inefficient, as a high `k` is wasteful for sequences with low acceptance rates, while a low `k` misses opportunities on highly predictable sequences.

This implementation introduces a `DynamicProposer` that monitors the historical acceptance rate for each request. It then dynamically increases or decreases `k` to keep the rate around a user-configurable target threshold. This aims to optimize throughput by tailoring the speculation length to the difficulty of the content being generated.

**Key Changes:**

  - Adds a new speculative method: `eagle_dynamic`.
  - Implements `DynamicProposer` which adjusts `k` per-sequence.
  - Introduces a new engine argument, `--acceptance-rate-threshold` to allow users to tune the target acceptance rate.
  - The data pipeline between the `GPUModelRunner` and the `Scheduler` has been updated to correctly handle variable-length draft proposals, ensuring accurate statistics and stable operation even when `k` differs between sequences in the same batch.

## Test Plan

The feature was tested end-to-end to observe the behavior in a live environment by running the vLLM engine with the `eagle_dynamic` method enabled. The dynamic adjustment of `k` can be observed in the logs.

**Command to run an E2E test:**

```bash
# The following command reflects the models used in testing.
python -m vllm.entrypoints.openai.api_server \
    --model meta-llama/Llama-3.1-8B-Instruct \
    --speculative-config '{"method": "eagle_dynamic", "model": "yuhuili/EAGLE3-LLaMA3.1-Instruct-8B", "num_speculative_tokens": 5, "acceptance_rate_threshold": 0.3}' \
    --max-model-len 4096
```

## Test Result

End-to-end testing confirms that the `k` value (the denominator in `accepted/k`) adapts dynamically and independently for each request based on its performance.

The logs below show `req=7` increasing its `k` from 9 to 10 due to high acceptance, while `req=5` decreases its `k` from 8 to 7 after a poor acceptance step.

```log
# k values are different for each request (req=5 has k=8, req=7 has k=9)
INFO [gpu_model_runner.py:949] EAGLE acceptance step: req=5 accepted=1/8 emitted=[123977, 627] | req=7 accepted=3/9 emitted=[11879, 374, 264, 27626]
INFO [gpu_model_runner.py:949] EAGLE acceptance step: req=5 accepted=1/8 emitted=[9, 116453] | req=7 accepted=3/9 emitted=[430, 1524, 279, 25655]

# req=7 has a good acceptance rate, so its k increases from 9 to 10
INFO [gpu_model_runner.py:949] EAGLE acceptance step: req=5 accepted=0/8 emitted=[118917] | req=7 accepted=4/10 emitted=[6299, 649, 1304, 264, 2466]

# req=5 has a poor acceptance rate, so its k decreases from 8 to 7
INFO [gpu_model_runner.py:949] EAGLE acceptance step: req=5 accepted=3/7 emitted=[114564, 104554, 18918, 86422] | req=7 accepted=3/10 emitted=[6811, 11, 323, 430]
```
---
<details>
<summary> Essential Elements of an Effective PR Description Checklist </summary>

- [x] The purpose of the PR, such as "Fix some issue (link existing issues this PR will resolve)".
- [x] The test plan, such as providing test command.
- [x] The test results, such as pasting the results comparison before and after, or e2e results
- [ ] (Optional) The necessary documentation update, such as updating `supported_models.md` and `examples` for a new model.
- [ ] (Optional) Release notes update. If your change is user facing, please update the release notes draft in the [Google Doc](https://docs.google.com/document/d/1YyVqrgX4gHTtrstbq8oWUImOyPCKSGnJ7xtTpmXzlRs/edit?tab=t.0).
</details>

### CODE REVIEW

**Comment by yang926** [comment]
Thanks for the careful follow up. It helps a lot.

For the first point, you’re right that in the current dynamic path `DynamicProposer.super().propose()` still runs the existing EAGLE pipeline and produces a `[B, max_k]` tensor, while `gpu_model_runner.propose_draft_token_ids` immediately uses `_true_draft_lengths` to restore ragged shapes.
The verifier then only relies on `draft_token_ids` + `cu_num_draft_tokens` (same pattern as n-gram), so the rectangular form isn’t strictly required by the verifier. In this PR I kept it mostly to avoid touching the EAGLE internals too much, but I agree it makes more sense for the dynamic EAGLE path to mirror n-gram more closely (emit drafts already truncated to per-seq `k` and treat them as ragged at the boundary).
Maybe I’ll rework that part accordingly.

For `cleanup_finished_seqs`: the intent was to drop `SequenceState` only for requests that are actually finished, not just temporarily unscheduled in a step. As you pointed out, with the current logic there is a risk of removing state for a swapped-out-but-still-alive request.
I’ll tighten this so that this code keep state for any request the scheduler still considers active and only delete entries when the engine marks the request as completed/removed.

Thanks again for pointing these out.

**Comment by fluctlux** [comment]
> Thanks for the careful follow up. It helps a lot.
> 
> For the first point, you’re right that in the current dynamic path `DynamicProposer.super().propose()` still runs the existing EAGLE pipeline and produces a `[B, max_k]` tensor, while `gpu_model_runner.propose_draft_token_ids` immediately uses `_true_draft_lengths` to restore ragged shapes. The verifier then only relies on `draft_token_ids` + `cu_num_draft_tokens` (same pattern as n-gram), so the rectangular form isn’t strictly required by the verifier. In this PR I kept it mostly to avoid touching the EAGLE internals too much, but I agree it makes more sense for the dynamic EAGLE path to mirror n-gram more closely (emit drafts already truncated to per-seq `k` and treat them as ragged at the boundary). Maybe I’ll rework that part accordingly.
> 
> For `cleanup_finished_seqs`: the intent was to drop `SequenceState` only for requests that are actually finished, not just temporarily unscheduled in a step. As you pointed out, with the current logic there is a risk of removing state for a swapped-out-but-still-alive request. I’ll tighten this so that this code keep state for any request the scheduler still considers active and only delete entries when the engine marks the request as completed/removed.
> 
> Thanks again for pointing these out.

Thank you for your reply, this answers my question.

**Comment by KookiesNKareem** [comment]
Glad to see per-request dynamic k getting built, the variable-length proposal plumbing here is exactly what any adaptive policy needs. I recently benchmarked this PR's policy shape (EWMA acceptance with a target band) and want to share data that might be useful for the signal design, plus a related gap.

Setup: Qwen2.5-7B FP16 target with an AWQ copy of itself as the draft, A40, ShareGPT prompts, greedy, prototyped on vLLM 0.6.3 (harness, controller, and raw logs: https://github.com/KookiesNKareem/spec-goodput-control).

The acceptance-threshold signal struggled on this workload: acceptance is nearly flat across load (0.86-0.91 from batch 1 to 32) while the right action flips from +23% (speculate at b=1) to -16% (disable at b=32). I ran acceptance-targeting at two setpoints against picking k by measured goodput (output tokens / wall-time per finished request, probing k in {1..4} plus off):

| batch | acc-target 0.80 | acc-target 0.90 | goodput | target-only |
|---|---|---|---|---|
| 1 | 49.2 tok/s | 46.4 | **56.4** (k4) | 33.8 |
| 32 | 682.7 (k1) | 735.0 (k1) | **715.1** (off) | **867.8** |

Neither setpoint works at both ends: acceptance dips at wide k, so they never hold k=4 at b=1, and they can't disable at b=32 where acceptance still reads a healthy 0.83. The goodput probe found k=4 at b=1 (+67%), which my own offline sweep had missed. So it might be worth considering measured goodput (or utility, cf. #44506) as the adaptation signal on top of this PR's plumbing.

The related gap: "off" isn't free. Even with the proposer fully skipped, a spec-configured engine loses 15-18% vs a clean engine at b=32. Same check on main (0.22.1): 1081 tok/s clean vs 920 with an ngram spec config attached (-15%), and V1 currently has no runtime off at all. Until off is a real action with "off = target-only" as the bar, that caps every adaptive-k policy, including this one.

Happy to help test this PR, or to prototype a goodput-based signal on top of it.

File: vllm/engine/arg_utils.py [hunk 1]
```diff
@@ -1350,6 +1357,8 @@ def create_load_config(self) -> LoadConfig:
             pt_load_map_location=self.pt_load_map_location,
         )
 
+        return SpeculativeConfig(**self.speculative_config)
+
```
Comment (hmellor) [inline vllm/engine/arg_utils.py:1361]: Why?

```suggestion
```

File: vllm/engine/arg_utils.py [hunk 2]
```diff
@@ -510,6 +510,7 @@ class EngineArgs:
     logits_processor_pattern: str | None = ModelConfig.logits_processor_pattern
 
     speculative_config: dict[str, Any] | None = None
+    acceptance_rate_threshold: float | None = None
```
Comment (hmellor) [inline vllm/engine/arg_utils.py:513]: Not necessary

```suggestion
```

File: vllm/engine/arg_utils.py [hunk 3]
```diff
@@ -1194,6 +1195,12 @@ def add_cli_args(parser: FlexibleArgumentParser) -> FlexibleArgumentParser:
         vllm_group.add_argument(
             "--speculative-config", **vllm_kwargs["speculative_config"]
         )
+        vllm_group.add_argument(
+            "--acceptance-rate-threshold",
+            type=float,
+            default=SpeculativeConfig.acceptance_rate_threshold,
+            help=SpeculativeConfig.acceptance_rate_threshold.__doc__,
+        )
```
Comment (hmellor) [inline vllm/engine/arg_utils.py:1203]: Not necessary

```suggestion
```

File: vllm/engine/arg_utils.py [hunk 4]
```diff
@@ -1366,6 +1375,9 @@ def create_speculative_config(
         if self.speculative_config is None:
             return None
 
+        if self.speculative_config is None:
+            self.speculative_config = {}
+
```
Comment (hmellor) [inline vllm/engine/arg_utils.py:1380]: Never reached

```suggestion
```

File: vllm/engine/arg_utils.py [hunk 5]
```diff
@@ -1375,6 +1387,12 @@ def create_speculative_config(
                 "target_parallel_config": target_parallel_config,
             }
         )
+
+        if self.acceptance_rate_threshold is not None:
+            self.speculative_config["acceptance_rate_threshold"] = (
+                self.acceptance_rate_threshold
+            )
+
```
Comment (hmellor) [inline vllm/engine/arg_utils.py:1395]: Not necessary

```suggestion
```

File: vllm/config/speculative.py [hunk 6]
```diff
@@ -441,9 +446,16 @@ def __post_init__(self):
                     ):
                         pass
                     else:
+                        eagle_method = self.method
+                        if eagle_method == "eagle_dynamic":
+                            if "eagle3" in self.draft_model_config.model.lower():
+                                eagle_method = "eagle3"
+                            else:
+                                eagle_method = "eagle"
```
Comment (hmellor) [inline vllm/config/speculative.py:454]: Why are we adding a new method if it just redirects to the other 2 types? Is it not a real type?

**Review by hmellor — CHANGES_REQUESTED** [review]
Blocking as there are a lot of basic Python issues in the simple parts of this PR

**Comment by hmellor** [comment]
Closing as stale
